### PR TITLE
Implement file channel based journal for Raft log

### PR DESCRIPTION
**Pull request by kuujo**
This PR replaces the existing journal with a new, more efficient `FileChannel` based journal. The `FileChannel` journal uses `Namespace`s to serialize/deserialize objects from the journal, thus avoiding an unnecessary memory copy when using memory mapped files. This PR also deprecates the use of the `MEMORY` storage level.

### CODE REVIEW

File: storage/src/main/java/io/atomix/storage/journal/FileChannelJournalSegmentReader.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  }
+
+  @Override
+  public Indexed<E> getCurrentEntry() {
+    return currentEntry;
+  }
+
+  @Override
+  public long getNextIndex() {
+    return currentEntry != null ? currentEntry.index() + 1 : firstIndex;
```
Comment (johnou) [inline storage/src/main/java/io/atomix/storage/journal/FileChannelJournalSegmentReader.java]: are entries guaranteed to be contiguous? why aren't we using nextEntry here?

File: storage/src/main/java/io/atomix/storage/journal/FileChannelJournalSegmentWriter.java [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+    lastEntry = null;
+
+    try {
+      // Truncate the index.
+      this.index.truncate(index);
+
+      if (index < segment.index()) {
+        channel.position(JournalSegmentDescriptor.BYTES);
+        memory.clear();
+        for (int i = 0; i < memory.limit(); i++) {
```
Comment (johnou) [inline storage/src/main/java/io/atomix/storage/journal/FileChannelJournalSegmentWriter.java]: utility method?

File: storage/src/main/java/io/atomix/storage/journal/SegmentedJournal.java [hunk 3]
```diff
@@ -369,115 +391,82 @@ synchronized void removeSegment(JournalSegment segment) {
...
-    Buffer buffer = HeapBuffer.allocate(descriptor.maxSegmentSize(), descriptor.maxSegmentSize());
+    ByteBuffer buffer = ByteBuffer.allocate(JournalSegmentDescriptor.BYTES);
     descriptor.copyTo(buffer);
+    buffer.flip();
+    try {
+      channel.write(buffer);
+    } catch (IOException e) {
+      throw new StorageException(e);
+    } finally {
+      try {
```
Comment (johnou) [inline storage/src/main/java/io/atomix/storage/journal/SegmentedJournal.java]: release `buffer` too?

Comment (kuujo) [reply storage/src/main/java/io/atomix/storage/journal/SegmentedJournal.java]: It's a `ByteBuffer`

File: storage/src/main/java/io/atomix/storage/journal/FileChannelJournalSegmentReader.java (hunk 1 above)
Comment (kuujo) [reply storage/src/main/java/io/atomix/storage/journal/FileChannelJournalSegmentReader.java]: Yes they're guaranteed to be contiguous. `nextEntry` gets its index from `getNextIndex()`, so I think there might be some weird problems with using it.

**Pull request merged**